### PR TITLE
[JBPM-9774] Incorrect object type returned in a DMN business-rule task of a BPMN

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9774

### CODE REVIEW

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/task/BusinessRuleTaskTest.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * Business rules task test. testing execution of rules with specified rule-flow group.
+ */
+public class BusinessRuleTaskTest extends JbpmTestCase {
+
+    private static final String RULE_TASK = "org/jbpm/test/functional/task/businessRuleTaskProcess.bpmn2";
+    private static final String RULE_TASK_ID = "org.jbpm.test.functional.task.businessRuleTask";
+
+    private static final String RULE_TASK_DRL = "org/jbpm/test/functional/task/RuleTask.drl";
```
Comment (tarilabs) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/task/BusinessRuleTaskTest.java:38]: unused?

Comment (tarilabs) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/task/BusinessRuleTaskTest.java:38]: unused?

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/task/buildPersonDecision.dmn [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		<dmn:itemComponent id="_A8E45253-E12B-4CB8-80EE-BB2DACDA8C6C" name="age" isCollection="false">
+			<dmn:typeRef>number</dmn:typeRef>
+		</dmn:itemComponent>
+		<dmn:itemComponent id="_8097D6E0-01DD-410D-A189-8D69704654F2" name="name" isCollection="false">
+			<dmn:typeRef>string</dmn:typeRef>
+		</dmn:itemComponent>
+	</dmn:itemDefinition>
+	<dmn:decision id="_F3595CA0-6666-4CF3-A81A-D21870D95379" name="buildPersonDecision">
+		<dmn:extensionElements/>
+		<dmn:variable id="_828DED89-ADA3-4973-8469-95C564DF06BD" name="personBRL" typeRef="Person"/>
```
Comment (tarilabs) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/task/buildPersonDecision.dmn:21]: I'm not sure how you ended up with the DMN model content, but it's a bit strange to me that variable and Decision are not sharing the same name?
Normally the editor would assign it the same name
can you confirm?

**Review by tarilabs — COMMENTED** [review]
some questions, please?

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/task/buildPersonDecision.dmn (hunk 2 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/task/buildPersonDecision.dmn:21]: I notice that the name of the outcome is the name of the variable and not the name of the decision. So for testing this properly I set them with different names.

Comment (tarilabs) [reply jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/task/buildPersonDecision.dmn:21]: I don't see why you should
DMN expects usually 
for the Decision name = Variable name
every DRGElement name = inherent variable name

**Review by tarilabs — COMMENTED** [review]
not resolved.
https://github.com/kiegroup/jbpm/pull/1947#discussion_r647405984
please leave the editor usage standard.

**Review by tarilabs — APPROVED** [review]
looks good to me, thank you @elguardian ! 🚀

**Review by jomarko — COMMENTED** [review]
@elguardian thank you for the PR. It seems to be working fine. I found just a single issue related to years and moths duration/java.time.Period.

I was not able to successfully map `years and months duration` from DMN into `java.time.Period` in BPMN/Java.

See commented parts of my PR https://github.com/elguardian/jbpm/pull/6

Is that something I should report separately, or I just selected wrong BPMN/Java type for `years and months duration`?

**Comment by tarilabs** [comment]
re https://github.com/kiegroup/jbpm/pull/1947#pullrequestreview-681623349

maybe just needs/missing something along the lines of

https://github.com/kiegroup/kogito-runtimes/blob/caa2a93186d856dbdb53047c9289a54568d0d7ec/kogito-codegen-modules/kogito-codegen-decisions/src/main/resources/class-templates/DecisionRestResourceQuarkusTemplate.java#L44-L46

?

**Comment by jomarko** [comment]
@tarilabs thank you, that works perfect for object serializing!

See:
- https://github.com/elguardian/jbpm/pull/6/commits/7071687afde7021c0ce944875fe90b54f6656ea4
- https://github.com/kiegroup/drools/pull/3669

However the `testBuildDMNPeriodProcess`  test, where single value is returned, returns `org.kie.dmn.feel.lang.types.impl.ComparablePeriod`, not a `java.time.Period`

What is your opinion about it. It is a separate issue, as it not anymore about returned object type?

**Comment by tarilabs** [comment]
I don't understand the question @jomarko , sorry?

When a DMN model returns a ComparablePeriod as part of the decision results, this would allow the underlying Jackson to serialize it as a usual Period in the ISO format. I believe once deserialized by Jackson to the Java Pojo type of the process variable, the Pojo would have the Period as a type of the field.

**Comment by jomarko** [comment]
@tarilabs ok, let me try put it differently.

```
ComparablePeriod variable = (ComparablePeriod) processInstance.getVariable("period");
```

works, while
```
Period variable = (Period) processInstance.getVariable("period");
```

Throws an error:
```
java.lang.ClassCastException: class org.kie.dmn.feel.lang.types.impl.ComparablePeriod cannot be cast to class java.time.Period (org.kie.dmn.feel.lang.types.impl.ComparablePeriod is in unnamed module of loader 'app'; java.time.Period is in module java.base of loader 'bootstrap')
```

What is for me hard to decide if:
- it is an issue
- it is expected behavior

**Comment by tarilabs** [comment]
I would expect a process variable would be of type `Period`, and not using internal classes of DMN (ComparablePeriod).

**Comment by elguardian** [comment]
@jomarko the info about deserialization is related to the java class dependant on how you map those clases so @tarilabs is right

**Comment by jomarko** [comment]
@elguardian hello, sorry, this PR was not ready for merge from my point of view. There is no ack from QE side. Also please notice I prepared PR where I fix some cases of mapping DMN types indo BPMN/Java types. https://github.com/kiegroup/jbpm/pull/1947#issuecomment-859514837

Without addressing them or reporting them separately, we can not mark JBPM-9774 as done.

**Comment by elguardian** [comment]
@jomarko provide a proper test plz. The test you provided is not using the new code. For the code make it work you need to provide class loader and runtime manager.

**Comment by jomarko** [comment]
@elguardian working on tests updates as https://github.com/kiegroup/jbpm/pull/1952

However I would like to ask update your added tests. I checked the newly created asset `businessRuleCollectionTaskProcess.bpmn2`, and it can be opened in our latest VS Code tooling. See the screenshot below. I think it is importnat we use assets that are compliant with current tooling.

Screenshot from 2021-06-14 14-20-13

**Comment by elguardian** [comment]
@jomarko This was requested by DMN support for generic collections. the problem is in stunner in this case. please file a jira to stunner team so they can allow this sort of mapping.

**Comment by jomarko** [comment]
@elguardian I got the same for `businessRuleTaskProcess.bpmn2`. Let me check my understanding. Does a JBPM-9744 brings a support for a DMN <-> BPMN mapping, that can not be modeled using our current tooling? Just manually? 

Screenshot from 2021-06-15 05-46-53

**Pull request merged**